### ISSUE TITLE
Selected Options cannot be sorted on creation

### ISSUE THREAD

**Issue by wmerrell** [yanickrochon/jquery.uix.multiselect#11]
I am trying to use your multiselect widget to select and sort some data on my Rails site. It seems to work and do what I want except that the selected items always appear in the same order. That is, I can use the widget to select and rearrange the items, and when I submit the form I get the selected items in the sorted order. However, if I revisit the page the same items are selected, but they are in the order of the available items list, not the order I selected on the previous visit. I have not found any way to change this. I did look at the sortMethod option, but it does not work properly. Even when I altered the code so the custom sort function was called properly, it did not change the order of the items. I think that this is because the sort function is being applied to the entire available items list and not just the selected items list. Any help would be appreciated.

**Comment by wmerrell** [comment]
Ok, I figured out how to make the selected list appear in the right order, by listing them in the desired order at the top of the available list. But now I am finding that when I sort the selected items and submit the form, submitted array is in the original order, not in the new sorted order. You can see this on your demo page. Oddly, on the demo page if I toggle the original element and back again, and then submit, then the submitted array is in the right order.

**Comment by yanickrochon** [comment]
You are right. The `sortMethod` option was originally defined with a default value, but it was changed a few commits back because this is not a general desired behaviour.

The default options do not sort any item at all by default and the options are listed in the order they appear in the original element. If you wish to apply a custom sort order, you might want to provide your own function (yes, the `sortMethod` accept a function receiving `a` and `b`, just like normal comparator). With this, you could provide some `data-selected-index` attribute or something. Or the server could pre-order the items so they would appear in the user's selection order (they _are_ inserted in the lists as they are loaded... or technically they should!).

Of course the widget won't remember the previous order, unless you set it up so it can know this information! Actually, the way it works right now _is_ by re-arranging the options in the original element so they are sorted the way the user wants them to be submitted. Just replace them in the same way, or provide some data and custom sorting function to define this order.

As or right now, this doesn't seem be a bug to me, unless I missed something?

Good luck!

**Comment by wmerrell** [comment]
Well, the sort function does not seem to do anything right now, so something seems like it needs fixing there. For my current purposes I don't need it, it works better to simply control the list supplied from the server and that works fine for now. 

But the rearranging of the selected items seems like it is broken. Go to your demo site, hit submit and look at the list that is returned. Now rearrange something and hit submit again. When I do this, I get the same list in the same order. If I cant get the new order, I may as well use a pick list. 

Thanks for getting back to me.

**Comment by yanickrochon** [comment]
Oh! Now _that_ is a problem! I forgot to reorder the selected element when the content of a group is re-ordered. I will try to push a fix for that as soon as possible.

Meanwhile, I'm not sure what you mean by "the sort function does not seem to do anything right now"? If you don't specify a sort method (a function or a predefined sort method), then there will be no sorting at all. Could you be more specific?

**Comment by wmerrell** [comment]
I wont swear that there is anything wrong, I may have jumped the gun on that. I was trying to sort only the selected list. I had two lists, one a list of the available options, and the other a list of the ones selected. When I tried to get this working, I dug into the sorting code and found some stuff around line 1037 in the reIndex function where it dealt with groups. My lists were not grouped, yet the code seemed to try to go down the group leg of the code. In short, my custom function was called with empty data for every item in the available list. I commented out some of it and got it calling my function with valid data, but since the approach itself was not right, I gave up on it and stopped using the sort approach. It seemed like it was not doing the right thing with non grouped lists, but I gave up before I can say for sure. Hope that helps.

**Comment by wmerrell** [comment]
I think it might be useful to mention something in the documentation about how to control the order of the selected items. Here is what I have done: On the server side I keep two lists. One is the list of all available options, and the other is the list of selected options kept in the order I desire. I then generate the list that will be used to create the widget with this formula: option_list = selected_list + (available_list - selected_list)

**Comment by yanickrochon** [comment]
Noted. I'll try to come up with an example for that use case, where you can provide some way to reload the page with the same list order.

**Comment by yanickrochon** [comment]
I will report this to a future release (version 2.1+), as it won't make it to version 2.0  I'd need to have more people wanting this automation to make it an official feature. As far as I understand it, implementing a custom sorting method and binding it to the `sortMethod` option would solve this issue without adding more overhead unto the widget codebase.
